torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,401 @@
+from ..modules.container import ModuleDict
+from ..modules.module import Module
+from ..parameter import Parameter
+from contextlib import contextmanager
+
+
+
+_cache_enabled = 0
+_cache = {}
+
+
+@contextmanager
+def cached():
+    r"""Context-manager that enables the caching system within
+    :class:`torch.nn.Parametrization`
+
+    This is usful when one uses certain parametrized parameter more than
+    once. An example of this is the loop in an RNN model
+    """
+    _cache_enabled += 1
+    try:
+        yield
+    finally:
+        _cache_enabled -= 1
+        if not _cache_enabled:
+            _cache = dict.fromkeys(_cache, None)
+
+
+class Parametrization(Module):
+    r"""A kind of Module that parametrizes a Parameter or a buffer in terms
+    of its forward function
+
+    After registering a Parametrization on a tensor ``t`` within a module with :meth:`
+    #register_parametrization`, ``module.t`` will be turn into a property, and will
+    return ``parametrization(t)``, where ``t`` is the previous tensor
+
+    After registering a Parametrization, the parametrization will be accessible under
+    ``module.parametrizations.t``. The unparametrized tensor may be accessed through
+    ``parametrization.original_tensor()``
+
+    Before a parametrization is registered, it may be chained to others to compose them
+    via the ``chain`` method.
+
+    Parametrizations may be registered on parameters, buffers, and already parametrized
+    tensors
+
+        # Defining several parametrizations on the same tensor
+        class MyParametrization(nn.Parametrization):
+            def forward(self, X):
+                return 2. * X
+        module = nn.Linear(5,3)
+        p1 = MyParametrization()
+        p2 = MyParametrization()
+        # Register p1 on the Parameter "weight"
+        register_parametrization(module, "weight", p1)
+        # Register p2 on the parametrized Parameter "weight"
+        register_parametrization(module, "weight", p2)
+        assert(module.weight == 4. * module.parametrizations.weight.original_tensor())))
+
+    We can also first compose them and then register them. The following code is
+    equivalent to the previous one
+
+        module = nn.Linear(5, 3)
+        # Chain the two parametrizations and register them
+        p2.chain(p1)
+        register_parametrization(module, "weight", p2)
+        # Now p2.evaluate() computes p1 and then p2
+        assert(module.weight == 4. * module.parametrizations.weight.original_tensor())))
+
+
+    Parametrized parameters and buffers have an in-built caching system via the context
+    manager :class:``cached``
+    """
+
+    def evaluate(self):
+        r"""Evaluates the parametrization"""
+        if self.is_registered():
+            return self(self.original)
+        else:
+            raise RuntimeError(
+                "A Parametrization cannot be evaluated before " "registering it"
```

Comment:
Can't we make accessing `.original` throw an error like that? Then we wouldn't need `evaluate` at all